MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c

```diff
@@ -818,19 +1119,82 @@ NvmeControllerInit (
-  Private->SqTdbl[0].Sqt = 0;
-  Private->SqTdbl[1].Sqt = 0;
-  Private->SqTdbl[2].Sqt = 0;
-  Private->CqHdbl[0].Cqh = 0;
-  Private->CqHdbl[1].Cqh = 0;
-  Private->CqHdbl[2].Cqh = 0;
-  Private->AsyncSqHead   = 0;
+  // MU_CHANGE [BEGIN] - Allocate IO Queue Buffer
+  //
+  // Set the Admin Queue Atttributes
+  // Set the sizes of the admin submission & completion queues in number of entries
+  //
+  // MU_CHANGE [BEGIN] - Support alternative hardware queue sizes in NVME driver
+  Aqa.Asqs  = PcdGetBool (PcdSupportAlternativeQueueSize) ? MIN (NVME_ALTERNATIVE_MAX_QUEUE_SIZE, Private->Cap.Mqes) : MIN (NVME_ASQ_SIZE, Private->Cap.Mqes);
+  Aqa.Rsvd1 = 0;
+  Aqa.Acqs  = PcdGetBool (PcdSupportAlternativeQueueSize) ? MIN (NVME_ALTERNATIVE_MAX_QUEUE_SIZE, Private->Cap.Mqes) : MIN (NVME_ACQ_SIZE, Private->Cap.Mqes);
+  Aqa.Rsvd2 = 0;
+  // MU_CHANGE [END] - Support alternative hardware queue sizes in NVME driver
+
+  //
+  // Save Queue Pair Data for admin queues in controller data structure
+  //
+  Private->SqData[0].NumberOfEntries = Aqa.Asqs;
+  Private->CqData[0].NumberOfEntries = Aqa.Acqs;
+
+  //
+  // Set admin queue entry size to default
+  //
+  Private->SqData[0].EntrySize = NVME_IOSQES_MIN;
+  Private->CqData[0].EntrySize = NVME_IOCQES_MIN;
+
+  // Calculate the number of pages required for the admin queues
+  AdminQueuePageCount = EFI_SIZE_TO_PAGES (Private->SqData[0].NumberOfEntries * (UINTN)LShiftU64 (2, Private->SqData[0].EntrySize))
+                        + EFI_SIZE_TO_PAGES (Private->CqData[0].NumberOfEntries * (UINTN)LShiftU64 (2, Private->CqData[0].EntrySize));
+  //
+  // Default:
+  // 6 x 4kB aligned buffers will be carved out of this buffer.
+  // 1st 4kB boundary is the start of the admin submission queue.
+  // 2nd 4kB boundary is the start of the admin completion queue.
+  // 3rd 4kB boundary is the start of I/O submission queue #1.
+  // 4th 4kB boundary is the start of I/O completion queue #1.
+  // 5th 4kB boundary is the start of I/O submission queue #2.
+  // 6th 4kB boundary is the start of I/O completion queue #2.
+  //
+  // Allocate 6 pages of memory, then map it for bus master read and write.
+  //
+  Status = PciIo->AllocateBuffer (
+                    PciIo,
+                    AllocateAnyPages,
+                    EfiBootServicesData,
+                    AdminQueuePageCount,
+                    (VOID **)&Private->Buffer,
+                    0
+                    );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Bytes  = EFI_PAGES_TO_SIZE (AdminQueuePageCount); // MU_CHANGE - Allocate IO Queue Buffer
+  Status = PciIo->Map (
+                    PciIo,
+                    EfiPciIoOperationBusMasterCommonBuffer,
+                    Private->Buffer,
+                    &Bytes,
+                    &MappedAddr,
+                    &Private->Mapping
+                    );
+  if (EFI_ERROR (Status) || (Bytes != EFI_PAGES_TO_SIZE (AdminQueuePageCount))) {
+    return Status;

```

Comment:
When returning here, do the Private->Buffer need to be freed? Or does the calling code know that it is responsible for freeing it?

Reply:
The calling code is responsible for freeing the buffer. In NvmExpress.c:NvmExpressDriverBindingStart, if this returns an error, it will goto exit and free. BUT if there isn't an error and Bytes != AdminQueuePages, it wont' return an error, will continue, but should reach the Exit label and free the buffer.
So I think we should return an error here regardless, to go straight to Exit on return from this function. `EFI_DEVICE_ERROR`?

Reply:
You are following what was done in NvmExpressDriverBindingStart, which seems fine.